internal/cucumber/tag/matcher.go

```diff
@@ -0,0 +1,62 @@
+// Package tag defines functions to parse cucumber feature files and filter them by cucumber tag expressions
+package tag
+
+import (
+	"io/fs"
+
+	gherkin "github.com/cucumber/gherkin/go/v28"
+	messages "github.com/cucumber/messages/go/v24"
+	tagexpressions "github.com/cucumber/tag-expressions/go/v6"
+	"github.com/rs/zerolog/log"
+)
+
+// MatchFiles finds feature files that include scenarios with tags that match the given tag expression.
+// A tag expression is a simple boolean expression including the logical operators "and", "or", "not".
+func MatchFiles(sys fs.FS, files []string, tagExpression string) (matched []string, unmatched []string) {
+	tagMatcher, err := tagexpressions.Parse(tagExpression)
+
+	if err != nil {
+		return matched, unmatched
+
+	}
+	for _, filename := range files {
+		f, err := sys.Open(filename)
+		if err != nil {
+			continue
+		}
+		defer f.Close()
+
+		uuid := &messages.UUID{}
```

Comment:
Imho, I'd move that out of the loop, as it doesn't need to be created per file. It also doesn't need to be a pointer, unlike the incremental ID generator which modifies itself. The UUID struct one is nothing but a wrapper around the real UUID package.

```suggestion
	}

	uuid := messages.UUID{}
	
	for _, filename := range files {
		f, err := sys.Open(filename)
		if err != nil {
			continue
		}
		defer f.Close()
```